### PR TITLE
Feature/output data keys

### LINKED ISSUE #764: Output dictionary keys hardcoded in model.py should be in DefaultDataKeys
https://github.com/PyTorchLightning/lightning-flash/blob/143112d01542083214d8d6b1bfba54688ac9cf3e/flash/core/model.py#L353-L355

It seems to me that hardcoding "loss" "y_hat" and "y" in here isn't the best practice and encourages everyone who needs those values in callbacks to either hardcode or store the key values in their own variables. For whatever reason, if the keys were to change, custom callbacks would break easily. 

I propose moving the output keys into the `DefaultDataKeys` for this use case. Naming them something with a prefix so as not to be confused with other input data keys. Alternatively, make a different enum class for these to go in and be used.

### PR DESCRIPTION
**Pull request by dlangerm**
## What does this PR do?

Removes hardcoded keys in the output of the core model. Creates new enumeration to track key names.

Fixes #764

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/core/model.py [hunk 1]
```diff
@@ -267,6 +268,20 @@ def __new__(mcs, *args, **kwargs):
         return result
 
 
+class DefaultOutputDataKeys(LightningEnum):
```
Comment (ethanwharris) [inline flash/core/model.py:271]:
```suggestion
class OutputKeys(LightningEnum):
```

File: flash/core/model.py [hunk 2]
```diff
@@ -331,11 +346,11 @@ def step(self, batch: Any, batch_idx: int, metrics: nn.ModuleDict) -> Any:
         x, y = batch
         y_hat = self(x)
         y, y_hat = self.apply_filtering(y, y_hat)
-        output = {"y_hat": y_hat}
-        y_hat = self.to_loss_format(output["y_hat"])
+        output = {DefaultOutputDataKeys.OUTPUT: y_hat}
+        y_hat = self.to_loss_format(output[DefaultOutputDataKeys.OUTPUT])
```
Comment (ethanwharris) [inline flash/core/model.py:350]:
```suggestion
        output = {OutputKeys.OUTPUT: y_hat}
        y_hat = self.to_loss_format(output[OutputKeys.OUTPUT])
```

File: flash/core/model.py [hunk 3]
```diff
@@ -331,11 +346,11 @@ def step(self, batch: Any, batch_idx: int, metrics: nn.ModuleDict) -> Any:
...
         losses = {name: l_fn(y_hat, y) for name, l_fn in self.loss_fn.items()}
 
-        y_hat = self.to_metrics_format(output["y_hat"])
+        y_hat = self.to_metrics_format(output[DefaultOutputDataKeys.OUTPUT])
```
Comment (ethanwharris) [inline flash/core/model.py:353]:
```suggestion
        y_hat = self.to_metrics_format(output[OutputKeys.OUTPUT])
```

File: flash/core/model.py [hunk 4]
```diff
@@ -350,9 +365,9 @@ def step(self, batch: Any, batch_idx: int, metrics: nn.ModuleDict) -> Any:
             logs["total_loss"] = sum(losses.values())
             return logs["total_loss"], logs
 
-        output["loss"] = self.compute_loss(losses)
-        output["logs"] = self.compute_logs(logs, losses)
-        output["y"] = y
+        output[DefaultOutputDataKeys.LOSS] = self.compute_loss(losses)
+        output[DefaultOutputDataKeys.LOGS] = self.compute_logs(logs, losses)
+        output[DefaultOutputDataKeys.TARGET] = y
```
Comment (ethanwharris) [inline flash/core/model.py:370]:
```suggestion
        output[OutputKeys.LOSS] = self.compute_loss(losses)
        output[OutputKeys.LOGS] = self.compute_logs(logs, losses)
        output[OutputKeys.TARGET] = y
```

File: flash/core/model.py [hunk 5]
```diff
@@ -380,16 +395,31 @@ def forward(self, x: Any) -> Any:
 
     def training_step(self, batch: Any, batch_idx: int) -> Any:
         output = self.step(batch, batch_idx, self.train_metrics)
-        self.log_dict({f"train_{k}": v for k, v in output["logs"].items()}, on_step=True, on_epoch=True, prog_bar=True)
-        return output["loss"]
+        self.log_dict(
+            {f"train_{k}": v for k, v in output[DefaultOutputDataKeys.LOGS].items()},
```
Comment (ethanwharris) [inline flash/core/model.py:399]:
```suggestion
            {f"train_{k}": v for k, v in output[OutputKeys.LOGS].items()},
```

File: flash/core/model.py [hunk 6]
```diff
@@ -380,16 +395,31 @@ def forward(self, x: Any) -> Any:
...
+            on_step=True,
+            on_epoch=True,
+            prog_bar=True,
+        )
+        return output[DefaultOutputDataKeys.LOSS]
```
Comment (ethanwharris) [inline flash/core/model.py:404]:
```suggestion
        return output[OutputKeys.LOSS]
```

File: flash/core/model.py [hunk 7]
```diff
@@ -380,16 +395,31 @@ def forward(self, x: Any) -> Any:
...
 
     def validation_step(self, batch: Any, batch_idx: int) -> None:
         output = self.step(batch, batch_idx, self.val_metrics)
-        self.log_dict({f"val_{k}": v for k, v in output["logs"].items()}, on_step=False, on_epoch=True, prog_bar=True)
+        self.log_dict(
+            {f"val_{k}": v for k, v in output[DefaultOutputDataKeys.LOGS].items()},
```
Comment (ethanwharris) [inline flash/core/model.py:409]:
```suggestion
            {f"val_{k}": v for k, v in output[OutputKeys.LOGS].items()},
```

File: flash/core/model.py [hunk 8]
```diff
@@ -380,16 +395,31 @@ def forward(self, x: Any) -> Any:
...
+            on_step=False,
+            on_epoch=True,
+            prog_bar=True,
+        )
 
     def test_step(self, batch: Any, batch_idx: int) -> None:
         output = self.step(batch, batch_idx, self.val_metrics)
-        self.log_dict({f"test_{k}": v for k, v in output["logs"].items()}, on_step=False, on_epoch=True, prog_bar=True)
+        self.log_dict(
+            {f"test_{k}": v for k, v in output[DefaultOutputDataKeys.LOGS].items()},
```
Comment (ethanwharris) [inline flash/core/model.py:418]:
```suggestion
            {f"test_{k}": v for k, v in output[OutputKeys.LOGS].items()},
```

File: flash/core/model.py [hunk 9]
```diff
@@ -267,6 +268,20 @@ def __new__(mcs, *args, **kwargs):
         return result
 
 
+class DefaultOutputDataKeys(LightningEnum):
+    """The ``DefaultDataKeys`` enum contains the keys that are used by built-in data sources to refer to inputs and
+    targets."""
```
Comment (ethanwharris) [inline flash/core/model.py:273]:
```suggestion
    """The ``OutputKeys`` enum contains the keys that are used internally by the ``Task`` when handling
    outputs."""
```

**Review by ethanwharris — APPROVED** [review]
Awesome, LGTM, just some name simplification :smiley:
